handlersettings.go

```diff
@@ -31,25 +53,71 @@ type protectedSettings struct {
-}
+// parseAndValidateSettings reads configuration from configFolder, decrypts it,
+// runs JSON-schema and logical validation on it and returns it back.
+func parseAndValidateSettings(ctx *log.Context, configFolder string) (h handlerSettings, _ error) {
+	ctx.Log("event", "reading configuration")
+	pubJSON, protJSON, err := readSettings(configFolder)
+	if err != nil {
+		return h, err
+	}
+	ctx.Log("event", "read configuration")
 
-// validate makes logical valiation on the handlerSettings which already passed
-// the schema validation.
-func (h handlerSettings) validate() error {
-	if h.publicSettings.CommandToExecute != "" &&
-		h.protectedSettings.CommandToExecute != "" {
-		return errCmdTooMany
+	ctx.Log("event", "validating json schema")
+	if err := validateSettingsSchema(pubJSON, protJSON); err != nil {
+		return h, errors.Wrap(err, "json validation error")
 	}
+	ctx.Log("event", "json schema valid")
 
-	if (h.protectedSettings.StorageAccountName != "") !=
-		(h.protectedSettings.StorageAccountKey != "") {
-		return errStoragePartialCredentials
+	ctx.Log("event", "parsing configuration json")
+	if err := vmextension.UnmarshalHandlerSettings(pubJSON, protJSON, &h.publicSettings, &h.protectedSettings); err != nil {
+		return h, errors.Wrap(err, "json parsing error")
 	}
+	ctx.Log("event", "parsed configuration json")
 
+	ctx.Log("event", "validating configuration logically")
+	if err := h.validate(); err != nil {
+		return h, errors.Wrap(err, "invalid configuration")
+	}
+	ctx.Log("event", "validated configuration")
+	return h, nil
+}
+
+// readSettings uses specified configFolder (comes from HandlerEnvironment) to
+// decrypt and parse the public/protected settings of the extension handler into
+// JSON objects.
+func readSettings(configFolder string) (pubSettingsJSON, protSettingsJSON map[string]interface{}, err error) {
+	pubSettingsJSON, protSettingsJSON, err = vmextension.ReadSettings(configFolder)
+	err = errors.Wrapf(err, "error reading extension configuration")
+	return
+}
+
+// validateSettings takes publicSettings and protectedSettings as JSON objects
+// and runs JSON schema validation on them.
+func validateSettingsSchema(pubSettingsJSON, protSettingsJSON map[string]interface{}) error {
+	pubJSON, err := toJSON(pubSettingsJSON)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal public settings into json")
+	}
+	protJSON, err := toJSON(protSettingsJSON)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal protected settings into json")
+	}
+
+	if err := validatePublicSettings(pubJSON); err != nil {
+		return err
+	}
+	if err := validateProtectedSettings(protJSON); err != nil {
+		return err
+	}
 	return nil
 }
+
+// toJSON converts given in-memory JSON object representation into a JSON object string.
+func toJSON(o map[string]interface{}) (string, error) {
+	if o == nil { // instead of JSON 'null' assume empty object '{}'
+		return "{}", nil
+	}
+	b, err := json.Marshal(o)
```

Comment:
@boumenot hmm this is just an internal operation and we pass the output to the json-schema validator. if we were to log it, it would be escaped with \t and \n s anyway into a single line due to the structured logger we use.